File: Hash Map - Set/1207 - UniqueNumberOfOccurrences.py
"""
1207. Unique Number of Occurrrences.

Given an array of integers arr, return true if the number of occurrences
of each value in the array is unique or false otherwise.
"""

class Solution(object):
    def uniqueOccurrences(self, arr):
        """
        :type arr: List[int]
        :rtype: bool
        """
        count_map = {}

        # Count occurrences
        for num in arr:
            if num in count_map:
                count_map[num] += 1
            else:
                count_map[num] = 1

        occurrence_set = set(count_map.values())
        return len(occurrence_set) == len(count_map)

s = Solution()

# Case 1
result1 = s.uniqueOccurrences(arr=[1,2,2,1,1,3])
expected1 = True
print(f"Result: {result1}, Expected: {expected1}")
assert result1 == expected1

# Case 2
result2 = s.uniqueOccurrences(arr=[1,2])
expected2 = False
print(f"Result: {result2}, Expected: {expected2}")
assert result2 == expected2

# Case 3
result3 = s.uniqueOccurrences(arr=[-3,0,1,-3,1,1,1,-3,10,0])
expected3 = True
print(f"Result: {result3}, Expected: {expected3}")
assert result3 == expected3

File: Hash Map & Set/1207_unique_number_of_occurrences.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <cassert>

class Solution{
    public:
    bool uniqueOccurrences(std::vector<int>& arr){
        std::unordered_map<int, int> count_map;

        for (int num : arr){
            if (count_map.find(num) != count_map.end()){
                count_map[num] += 1;
            }else{
                count_map[num] = 1;
            }
        }

        std::unordered_set<int> occurrence_set;
        for (const auto& entry: count_map){
            occurrence_set.insert(entry.second);
        }

        return occurrence_set.size() == count_map.size();
    }
};

int main()
{
    Solution s;

    // Case 1
    std::vector<int> arr = {1,2,2,1,1,3};
    bool result = s.uniqueOccurrences(arr);
    bool expected = true;
    std::cout << "Result: " << result << ", Expected: " << expected << std::endl;

    std::vector<int> arr2 = {1,2};
    bool result2 = s.uniqueOccurrences(arr2);
    bool expected2 = false;
    std::cout << "Result: " << result2 << ", Expected: " << expected2 << std::endl;

    std::vector<int> arr3 = {-3,0,1,-3,1,1,1,-3,10,0};
    bool result3 = s.uniqueOccurrences(arr3);
    bool expected3 = true;
    std::cout << "Result: " << result3 << ", Expected: " << expected3 << std::endl;
}